Changes in module 'nf-core/vcf2cytosure'
--- modules/nf-core/vcf2cytosure/main.nf
+++ modules/nf-core/vcf2cytosure/main.nf
@@ -12,7 +12,7 @@
     tuple val(meta2), path(coverage_bed)
     tuple val(meta3), path(cns)
     tuple val(meta4), path(snv_vcf)
-    path blacklist_bed
+    path(blacklist_bed)
 
     output:
     tuple val(meta), path("*.cgh"), emit: cgh
@@ -56,7 +56,7 @@
     def prefix = task.ext.prefix ?: "${meta.id}"
 
     """
-    touch ${prefix}.xml
+    touch ${prefix}.cgh
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
