Changes in module 'nf-core/gatk4/germlinecnvcaller'
--- modules/nf-core/gatk4/germlinecnvcaller/main.nf
+++ modules/nf-core/gatk4/germlinecnvcaller/main.nf
@@ -36,7 +36,7 @@
         avail_mem = (task.memory.mega*0.8).intValue()
     }
     """
-    gatk --java-options "-Xmx${avail_mem}g" GermlineCNVCaller \\
+    gatk --java-options "-Xmx${avail_mem}M" GermlineCNVCaller \\
         $input_list \\
         $ploidy_command \\
         $output_command \\

************************************************************
